_posts/2021-03-15-forecasting-time-series-with-torch_3/forecasting_3.Rmd

```diff
@@ -0,0 +1,445 @@
+
+The reason we do this is owed to the way we are going to train the network. With *seq2seq*, people often use a technique called "teacher forcing", where instead of feeding back its own prediction into the decoder module, you pass it the value it *should* have predicted. To be clear, this is done during training only, and to a configurable degree.
+
+```{r}
+n_timesteps <- 7 * 24 * 2
+n_forecast <- n_timesteps
+
+vic_elec_get_year <- function(year, month = NULL) {
+  vic_elec %>%
+    filter(year(Date) == year, month(Date) == if (is.null(month)) month(Date) else month) %>%
+    as_tibble() %>%
+    select(Demand)
+}
+
+elec_train <- vic_elec_get_year(2012) %>% as.matrix()
+elec_valid <- vic_elec_get_year(2013) %>% as.matrix()
+elec_test <- vic_elec_get_year(2014, 1) %>% as.matrix()
+
+train_mean <- mean(elec_train)
+train_sd <- sd(elec_train)
+
+elec_dataset <- dataset(
+  name = "elec_dataset",
+  
+  initialize = function(x, n_timesteps, sample_frac = 1) {
+    
+    self$n_timesteps <- n_timesteps
+    self$x <- torch_tensor((x - train_mean) / train_sd)
+    
+    n <- length(self$x) - self$n_timesteps - 1
+    
+    self$starts <- sort(sample.int(
+      n = n,
+      size = n * sample_frac
+    ))
+    
+  },
+  
+  .getitem = function(i) {
+    
+    start <- self$starts[i]
+    end <- start + self$n_timesteps - 1
+    lag <- 1
+    
+    list(
+      x = self$x[start:end],
+      y = self$x[(start+lag):(end+lag)]$squeeze(2)
+    )
+    
+  },
+  
+  .length = function() {
+    length(self$starts) 
+  }
+)
+```
+
+Dataset as well as dataloader instantations then can proceed as before.
+
+```{r}
+batch_size <- 32
+
+train_ds <- elec_dataset(elec_train, n_timesteps, sample_frac = 0.5)
+train_dl <- train_ds %>% dataloader(batch_size = batch_size, shuffle = TRUE)
+
+valid_ds <- elec_dataset(elec_valid, n_timesteps, sample_frac = 0.5)
+valid_dl <- valid_ds %>% dataloader(batch_size = batch_size)
+
+test_ds <- elec_dataset(elec_test, n_timesteps)
+test_dl <- test_ds %>% dataloader(batch_size = 1)
+```
+
+# Model
+
+Technically, the model consists of three *modules*: the aforementioned encoder and decoder, and the *seq2seq* module that orchestrates them.
+
+### Encoder
+
+The encoder takes its input and runs it through an RNN. Of the two things returned by a recurrent neural network, outputs and state, so far we've only been using the former. This time, we do the opposite: We throw away the outputs, and only return the state.
+
+If the RNN in question is a GRU (and assuming that of the outputs, we take just the final time step, which is what we've been doing throughout), there really is no difference: The final state equals the final output. If it's an LSTM, however, there is a second, the "cell", state. In that case, returning the state instead of the final output will carry more information.
```

Comment:
```suggestion
If the RNN in question is a GRU (and assuming that of the outputs, we take just the final time step, which is what we've been doing throughout), there really is no difference. The final state equals the final output. If it's an LSTM, however, there is a second state, the "cell". In that case, returning the state instead of the final output will carry more information.
```

Reply:
You've got a lot of the colon-to-new-sentence construct, so I'm paring that back a bit. Also, unless there are two "the cells," then the second sentence I modified doesn't make sense as is (though it should with the edit).

Reply:
hm, no, it really should be "cell state" ... I think I can do
> second kind of state, the "cell state"

Reply:
Yeah, sounds much better. Originally there was a comma between "cell" and state, so I didn't even put them together.

Reply:
haha - i see ... that was a comma of type "no idea whether it should go there or not" :-)